clients/src/main/java/org/apache/kafka/clients/consumer/internals/StreamsGroupHeartbeatRequestManager.java

```diff
@@ -0,0 +1,387 @@
+                repartitionTopicInfo.setName(repartitionTopic.getKey());
+                repartitionTopic.getValue().numPartitions().ifPresent(repartitionTopicInfo::setPartitions);
+                repartitionTopic.getValue().replicationFactor().ifPresent(repartitionTopicInfo::setReplicationFactor);
+                repartitionTopic.getValue().topicConfigs().forEach((k, v) ->
+                    repartitionTopicInfo.topicConfigs().add(new StreamsGroupHeartbeatRequestData.KeyValue().setKey(k).setValue(v))
+                );
+                repartitionTopicsInfo.add(repartitionTopicInfo);
+            }
+            repartitionTopicsInfo.sort(Comparator.comparing(StreamsGroupHeartbeatRequestData.TopicInfo::name));
+            return repartitionTopicsInfo;
+        }
+
+        private static List<StreamsGroupHeartbeatRequestData.TopicInfo> getChangelogTopicsInfoFromStreams(
+            final StreamsRebalanceData.Subtopology subtopologyDataFromStreams) {
+            final List<StreamsGroupHeartbeatRequestData.TopicInfo> changelogTopicsInfo = new ArrayList<>();
+            for (final Map.Entry<String, StreamsRebalanceData.TopicInfo> changelogTopic : subtopologyDataFromStreams.stateChangelogTopics().entrySet()) {
+                final StreamsGroupHeartbeatRequestData.TopicInfo changelogTopicInfo = new StreamsGroupHeartbeatRequestData.TopicInfo();
+                changelogTopicInfo.setName(changelogTopic.getKey());
+                changelogTopic.getValue().replicationFactor().ifPresent(changelogTopicInfo::setReplicationFactor);
+                changelogTopic.getValue().topicConfigs().forEach((k, v) ->
+                    changelogTopicInfo.topicConfigs().add(new StreamsGroupHeartbeatRequestData.KeyValue().setKey(k).setValue(v))
+                );
+                changelogTopicsInfo.add(changelogTopicInfo);
+            }
+            changelogTopicsInfo.sort(Comparator.comparing(StreamsGroupHeartbeatRequestData.TopicInfo::name));
+            return changelogTopicsInfo;
+        }
+    }
+
+
+    /**
+     * Represents the state of a heartbeat request, including logic for timing, retries, and exponential backoff. The object extends
+     * {@link RequestState} to enable exponential backoff and duplicated request handling. The two fields that it holds are:
+     */
+    static class HeartbeatRequestState extends RequestState {
+
+        /**
+         * heartbeatTimer tracks the time since the last heartbeat was sent
+         */
+        private final Timer heartbeatTimer;
+
+        /**
+         * The heartbeat interval which is acquired/updated through the heartbeat request
+         */
+        private long heartbeatIntervalMs;
+
+        public HeartbeatRequestState(
+            final LogContext logContext,
+            final Time time,
+            final long heartbeatIntervalMs,
+            final long retryBackoffMs,
+            final long retryBackoffMaxMs,
+            final double jitter) {
```

Comment:
This is the one parameter list formatting where I agree it's not nice due to the single indentation.